File: fnc/examples/chapter07/hellosvd.m
%%
% We make an image from some text, then reload it as a matrix.
tobj = text(0,0,'Hello world','fontsize',44);
ex = get(tobj,'extent');      % ignore this line 
axis([ex(1) ex(1)+ex(3) ex(2) ex(2)+ex(4)]), axis off   % ignore this line 
saveas(gcf,'hello.png')
A = imread('hello.png');
A = double(rgb2gray(A));
imagesc(A),  colormap gray
[m,n] = size(A)

%%
% Next we show that the singular values decrease exponentially, until they
% reach zero (more precisely, are about $\sigma_1 \macheps$). For all
% numerical purposes, this determines the rank of the matrix.
[U,S,V] = svd(A);
sigma = diag(S);
semilogy(sigma,'.')
title('singular values'), axis tight        % ignore this line 
xlabel('i'), ylabel('\sigma_i')  % ignore this line 
r = find(sigma/sigma(1) > 10*eps,1,'last')

%%
% The rapid decrease suggests that we can get fairly good low-rank
% approximations. 
for i = 1:4
    subplot(2,2,i)
    k = 2*i;
    Ak = U(:,1:k)*S(1:k,1:k)*V(:,1:k)';
    imshow(Ak,[0 255])
    title(sprintf('rank = %d',k))
end

%%
% Consider how little data is needed to reconstruct these images. For rank
% 8, for instance, we have 8 left and right singular vectors plus 8
% singular values, for a compression ratio of better than 25:1.  
compression = 8*(m+n+1) / (m*n)

File: julia/scripts/chapter07/hellosvd.jl

using Plots
plot([],[],leg=:none,annotations=(0.5,0.5,text("Hello world",44,:center,:middle)),
    grid=:none,frame=:none)

savefig("hello.png")

using Images
img = load("hello.png")
A = @. Float64(Gray(img));
@show m,n = size(A);

using LinearAlgebra
U,sigma,V = svd(A)

scatter(sigma,m=(:o,2),
    title="Singular values",xaxis=("\$i\$"), yaxis=(:log10,"\$\\sigma_i\$"),leg=:none )

r = findlast(@.sigma/sigma[1] > 10*eps())

Ak = [ U[:,1:k]*diagm(0=>sigma[1:k])*V[:,1:k]' for k=2*(1:4) ]
reshape( [ @.Gray(Ak[i]) for i=1:4 ],2,2)

compression = 8*(m+n+1) / (m*n)
